File: ts/Solution_19.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}
function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    const dummy = new ListNode(0, head);
    let fast: ListNode | null = dummy;
    let slow: ListNode | null = dummy;

    // 快指针先移动 n+1 步
    for (let i = 0; i <= n; i++) {
        if (fast === null) break;  // 边界保护
        fast = fast.next;
    }

    // 双指针同步移动直到快指针到末尾
    while (fast !== null) {
        fast = fast.next;
        slow = slow!.next;  // 非空断言（slow 不会为 null）
    }

    // 删除目标节点
    if (slow !== null && slow.next !== null) {
        slow.next = slow.next.next;
    }

    return dummy.next;
};

File: golang/Solution_19.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
  type ListNode struct {
      Val int
      Next *ListNode
  }
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy:=&ListNode(Next:head)
	fast,slow:=dummy,dummy
	for i:=0;i<=n;i++{
		if fast==nil{
			return head
		}
		fast=fast.Next
	}
	for fast!=nil{
		fast=fast.Next
		slow=slow.Next
	}
	if slow.Next!=nil{
		slow.Next=slow.Next.Next
	}
	return dummy.Next
}